nav2_collision_monitor/src/collision_monitor_node.cpp

```diff
@@ -434,34 +437,34 @@ void CollisionMonitor::process(const Velocity & cmd_vel_in, const std_msgs::msg:
         break;
       }
     }
-  }
 
-  if (collision_points_marker_pub_->get_subscription_count() > 0) {
-    // visualize collision points with markers
-    auto marker_array = std::make_unique<visualization_msgs::msg::MarkerArray>();
-    visualization_msgs::msg::Marker marker;
-    marker.header.frame_id = get_parameter("base_frame_id").as_string();
-    marker.header.stamp = rclcpp::Time(0, 0);
-    marker.ns = "collision_points";
-    marker.id = 0;
-    marker.type = visualization_msgs::msg::Marker::POINTS;
-    marker.action = visualization_msgs::msg::Marker::ADD;
-    marker.scale.x = 0.02;
-    marker.scale.y = 0.02;
-    marker.color.r = 1.0;
-    marker.color.a = 1.0;
-    marker.lifetime = rclcpp::Duration(0, 0);
-    marker.frame_locked = true;
-
-    for (const auto & point : collision_points) {
-      geometry_msgs::msg::Point p;
-      p.x = point.x;
-      p.y = point.y;
-      p.z = 0.0;
-      marker.points.push_back(p);
+    if (collision_points_marker_pub_->get_subscription_count() > 0) {
+      // visualize collision points with markers
+      auto marker_array = std::make_unique<visualization_msgs::msg::MarkerArray>();
+      visualization_msgs::msg::Marker marker;
+      marker.header.frame_id = get_parameter("base_frame_id").as_string();
+      marker.header.stamp = rclcpp::Time(0, 0);
+      marker.ns = "collision_points";
```

Comment:
Or perhaps have the `ns` be `"collision_points_" + source->getName()` so that they're namespaced separately? I forget how rviz handles `ns` versus `id` field for visualizing multiple messages

Reply:
The behavior in rviz wasn't good with any of the solutions (using `ns` or `id`). The makers seemed to be correctly published but they were displayed with a very slow rate in rviz. The delay between the sensors actual data display and the markers update was something like a 10sec... I think publishing a new message for each source on a single topic cannot work and is not a nice design.

A solution is to actually use the `visualization_msgs::msg::MarkerArray` type. If we add a marker to the `MarkerArray` message for each source, and publish this array only once when calling `process()`, markers display works great in rviz, with good update rate. Plus, if we add the source's name in `marker.ns`, we can easily select the sources we actually want to check in rviz "Display" panel.

![5_sources_rviz](https://github.com/user-attachments/assets/cb674cf1-f45a-4ec7-b20f-ca35c601d971)

Here is an example for 5 sources (a scan in dark blue and 4 ranges). I only added the different colors for this test, this change is not on PR branch.

Reply:
The solution I told you about is already commit on this PR branch. So currently this is the structure:
```c++
process()  {

 ... // some init

  auto marker_array = std::make_unique<visualization_msgs::msg::MarkerArray>();
  for (std::shared_ptr<Source> source : sources_) {

   ... // checking source is valid
   

    if (collision_points_marker_pub_->get_subscription_count() > 0) {
      visualization_msgs::msg::Marker marker;
      marker.ns = "collision_points_" + source->getSourceName();

      ... // filling marker data for current source
      
      marker_array->markers.push_back(marker);
    } // end of check on marker topic subscription count
  } // end of source loop

  // publishing one marker array message containing all sources (on marker in array per source)
  if (collision_points_marker_pub_->get_subscription_count() > 0) {
    collision_points_marker_pub_->publish(std::move(marker_array));
  }

  ... // checking polygons and actions, publishing velocity and polygons
} // end of process function
```

It allows the results on the rviz screenshot with normal update rate, but the color of all markers in array is red